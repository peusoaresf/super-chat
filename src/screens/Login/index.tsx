import React, { useEffect, useState } from 'react';
import { View, Text } from 'react-native';
import { StackNavigationProp } from '@react-navigation/stack';
import { signup, signin } from '../../services/authService'
import { getUserDetails } from '../../services/userService'
import { user as userAtom } from '../../atoms/user'
import { useSetRecoilState } from 'recoil'
import { config } from '../../services/request'

export default function Login({ navigation }: {
  navigation: StackNavigationProp<any, any>
}) {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [buttonDisabled, setButtonDisabled] = useState(false);
  const setUser = useSetRecoilState(userAtom)

  useEffect(() => {
    setButtonDisabled((!username || !password));
  }, [username, password])

  const handleUserLogin = async (getToken) => {
    const token = await getToken(username, password)
    config.authToken = token
    const userDetails = await getUserDetails()
    setUser(userDetails)
    navigation.navigate('Home')
  }

  return (
    <View style={styles.screenContainer}>
      <View style={styles.formContainer}>
        <UsernameInput
          value={username}
          onChangeText={(text) => setUsername(text)} />
        <PasswordInput
          value={password}
          onChangeText={(text) => setPassword(text)} />
        <LoginButton onPress={() => handleUserLogin(signin)} disabled={buttonDisabled} />
        <SignUpButton onPress={() => handleUserLogin(signup)} disabled={buttonDisabled} />
      </View>
    </View>
  );
}

import { TextInput, StyleSheet, GestureResponderEvent } from 'react-native';

function Input({ placeholder, value, onChangeText, secureTextEntry }: {
  placeholder: string,
  value: string,
  onChangeText: (a: string) => void,
  secureTextEntry?: boolean
}) {
  return (
    <TextInput
      autoCapitalize="none"
      style={styles.input}
      placeholder={placeholder}
      value={value}
      onChangeText={onChangeText}
      secureTextEntry={secureTextEntry} />
  );
}

function UsernameInput({ value, onChangeText }: {
  value: string,
  onChangeText: (a: string) => void
}) {
  return (
    <Input
      placeholder="Enter your username"
      value={value}
      onChangeText={onChangeText} />
  );
}

function PasswordInput({ value, onChangeText }: {
  value: string,
  onChangeText: (a: string) => void
}) {
  return (
    <Input
      placeholder="Enter your password"
      value={value}
      onChangeText={onChangeText}
      secureTextEntry={true} />
  );
}

import { TouchableHighlight } from 'react-native';

function Button({ text, onPress, disabled }: {
  text: string,
  onPress: (event: GestureResponderEvent) => void,
  disabled?: boolean
}) {
  return (
    <TouchableHighlight disabled={disabled} style={styles.button} onPress={onPress}>
      <Text style={!disabled ? styles.buttonText : { color: 'gray' }}>{text}</Text>
    </TouchableHighlight>
  );
}

function LoginButton({ onPress, disabled }: {
  onPress: (event: GestureResponderEvent) => void,
  disabled?: boolean
}) {
  return (
    <Button
      text="Login"
      onPress={onPress}
      disabled={disabled} />
  );
}

function SignUpButton({ onPress, disabled }: {
  onPress: (event: GestureResponderEvent) => void,
  disabled?: boolean
}) {
  return (
    <Button
      text="Sign Up"
      onPress={onPress}
      disabled={disabled} />
  );
}

const styles = StyleSheet.create({
  screenContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
    backgroundColor: '#E5DDD5'
  },
  formContainer: {
    height: 250,
    width: '100%',
    justifyContent: 'space-between'
  },
  input: {
    padding: 10,
    borderRadius: 10,
    width: '100%',
    backgroundColor: 'white'
  },
  button: {
    borderRadius: 10,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 10,
    backgroundColor: '#075E55'
  },
  buttonText: {
    color: 'white'
  }
});