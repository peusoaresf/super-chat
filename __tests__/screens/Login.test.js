import React from 'react';
import renderer from 'react-test-renderer';

import Login from '../../src/screens/Login.tsx';

jest.useFakeTimers();

it('renders correctly', () => {
  const tree = renderer.create(<Login />).toJSON();
  expect(tree).toMatchSnapshot();
});