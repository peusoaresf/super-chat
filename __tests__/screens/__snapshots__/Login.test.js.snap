// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E5DDD5",
      "flex": 1,
      "justifyContent": "center",
      "padding": 20,
    }
  }
>
  <View
    style={
      Object {
        "height": 250,
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Enter your username"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 10,
          "padding": 10,
          "width": "100%",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Enter your password"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 10,
          "padding": 10,
          "width": "100%",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#075E55",
          "borderRadius": 10,
          "justifyContent": "center",
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
          }
        }
      >
        Login
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#075E55",
          "borderRadius": 10,
          "justifyContent": "center",
          "padding": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
          }
        }
      >
        Sign Up
      </Text>
    </View>
  </View>
</View>
`;
